apps/aecore/lib/aecore/contract/tx/contract_call_tx.ex

```diff
@@ -0,0 +1,377 @@
+         encoded_sender,
+         nonce,
+         encoded_contract,
+         vm_version,
+         fee,
+         ttl,
+         amount,
+         gas,
+         gas_price,
+         call_data
+       ]) do
+     payload = %ContractCallTx{
+       contract: encoded_contract,
+       vm_version: vm_version,
+       amount: amount,
+       gas: gas,
+       gas_price: gas_price,
+       call_data: call_data
+     }
+
+     DataTx.init_binary(ContractCallTx, payload, [encoded_sender], fee, nonce, ttl)
+   end
+
+   @spec is_minimum_fee_met?(SignedTx.t()) :: boolean()
+   def is_minimum_fee_met?(tx) do
+     tx.data.fee >= Application.get_env(:aecore, :tx_data)[:minimum_fee]
+   end
+
+   defp run_contract(
+          %ContractCallTx{
+            contract: contract_address,
+            vm_version: vm_version,
+            amount: amount,
+            gas: gas,
+            gas_price: gas_price,
+            call_data: call_data,
+            call_stack: call_stack
+          },
+          %{caller_address: caller_address} = call,
+          block_height,
+          _nonce,
+          chain_state
+        ) do
+     contracts_tree = chain_state.contracts
+     # check the get_contract
+     contract = ContractStateTree.get_contract(contracts_tree, contract_address.value)
+
+     call_definition = %{
+       caller: caller_address,
+       contract: contract_address,
+       gas: gas,
+       gas_price: gas_price,
+       call_data: call_data,
+       amount: amount,
+       call_stack: call_stack,
+       code: contract.code,
+       call: call,
+       height: block_height
+     }
+
+     Dispatch.run(vm_version, call_definition, chain_state)
+   end
+
+   defp check_account_balance(accounts, sender, required_amount) do
+     case AccountStateTree.get(accounts, sender).balance - required_amount > 0 do
+       true -> :ok
+       false -> {:error, "#{__MODULE__}: Negative balance"}
+     end
+   end
+
+   defp check_contract_balance(accounts, sender, amount) do
+     case AccountStateTree.get(accounts, sender) do
+       %Account{} ->
+         check_validity(
+           Account.balance(accounts, sender) >= amount,
+           "#{__MODULE__}: Insufficient funds"
+         )
+
+       :none ->
+         {:error, "#{__MODULE__}: Contract not found"}
+     end
+   end
+
+   defp check_call(
+          %ContractCallTx{contract: contract, vm_version: vm_version},
+          chain_state
+        ) do
+     case ContractStateTree.get_contract(chain_state.contracts, contract.value) do
+       %Contract{} = contract ->
+         case contract.vm_version == vm_version do
+           true -> :ok
+           false -> {:error, "#{__MODULE__}: Wrong VM version"}
+         end
+
+       :none ->
+         {:error, "#{__MODULE__}: Contract does not exist"}
+     end
+   end
+
+   defp check_validity(true, _), do: :ok
+   defp check_validity(false, message), do: {:error, message}
+
+   defp validate_identifier(%Identifier{} = id, type) do
+     case type do
```

Comment:
This `type` matching doesn't seem to be needed. You can just `Identifier.create_identity(id.value, type) == id` instead